Common/Util/RateLimit/LeakyBucket.cs

```diff
@@ -0,0 +1,168 @@
+        /// of tokens</param>
+        public void Consume(long tokens, long timeout = Timeout.Infinite)
+        {
+            if (timeout < Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout),
+                    "Invalid timeout. Use -1 for no timeout, 0 for immediate timeout and a positive number " +
+                    "of milliseconds to indicate a timeout. All other values are out of range."
+                );
+            }
+
+            var startTime = _timeProvider.GetUtcNow();
+
+            while (true)
+            {
+                if (TryConsume(tokens))
+                {
+                    break;
+                }
+
+                if (timeout != Timeout.Infinite)
+                {
+                    // determine if the requested timeout has elapsed
+                    var currentTime = _timeProvider.GetUtcNow();
+                    var elapsedMilliseconds = (currentTime - startTime).TotalMilliseconds;
+                    if (elapsedMilliseconds > timeout)
+                    {
+                        throw new TimeoutException("The operation timed out while waiting for the rate limit to be lifted.");
+                    }
+                }
+
+                _sleep.Sleep();
+            }
+        }
+
+        /// <summary>
+        /// Attempts to consume the specified number of tokens from the bucket. If the
+        /// requested number of tokens are not immediately available, then this method
+        /// will return false to indicate that zero tokens have been consumed.
+        /// </summary>
+        public bool TryConsume(long tokens)
+        {
+            if (tokens <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(tokens),
+                    "Number of tokens to consume must be positive"
+                );
+            }
+
+            if (tokens > _capacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(tokens),
+                    "Number of tokens to consume must be less than or equal to the capacity"
+                );
+            }
+
+            lock (_sync)
+            {
+                // determine how many units have become available since last invocation
+                var refilled = Math.Max(0, _refill.Refill());
+
+                // the number of tokens to add, the max of which is the difference between capacity and currently available
+                var deltaTokens = Math.Min(_capacity - _available, refilled);
+
+                // update the available number of units with the new tokens
+                _available += deltaTokens;
+
+                if (tokens > _available)
+                {
+                    // we don't have enough tokens yet
+                    Logging.Log.Trace($"LeakyBucket.TryConsume({tokens}): Failed to consumed tokens. Available: {_available}");
```

Comment:
Should it be `tokens >= _available`, can we consume all available? Maybe this log would be better as an error log

Reply:
We can consume all of the tokens but this condition is checking the opposite. It's asking if there's not enough tokens (requested more than available). I don't view this as an error condition since it's in a `TryXXX` method and the consumer has enough information at its disposal to perform the same logging, if desired.